lib/engine/pa_messages.ex

```diff
@@ -48,47 +48,39 @@ defmodule Engine.PaMessages do
     {:noreply, state}
   end
 
-  defp send_pa_messages(pa_messages, pa_messages_last_sent) do
-    for %{
-          "id" => pa_id,
-          "visual_text" => visual_text,
-          "audio_text" => audio_text,
-          "interval_in_minutes" => interval_in_minutes,
-          "priority" => priority,
-          "sign_ids" => sign_ids
-        } <- pa_messages,
-        into: %{} do
-      active_pa_message = %PaMessage{
-        id: pa_id,
-        visual_text: visual_text,
-        audio_text: audio_text,
-        priority: priority,
-        sign_ids: sign_ids,
-        interval_in_ms: interval_in_minutes * @minute_in_ms
-      }
-
-      {_, last_sent_time} = Map.get(pa_messages_last_sent, pa_id, {nil, nil})
-
-      time_since_last_send =
-        if last_sent_time,
-          do: DateTime.diff(DateTime.utc_now(), last_sent_time, :millisecond),
-          else: active_pa_message.interval_in_ms
-
-      if time_since_last_send >= active_pa_message.interval_in_ms do
-        send_pa_message(active_pa_message)
-        {pa_id, {active_pa_message, DateTime.utc_now()}}
+  defp select_pa_messages(pa_messages, state) do
+    now = DateTime.utc_now()
+
+    Enum.flat_map_reduce(pa_messages, state, fn message, state ->
+      last_sent = Map.get(state.pa_messages_last_sent, message.id, DateTime.from_unix!(0))
+
+      if DateTime.diff(DateTime.utc_now(), last_sent, :millisecond) >= message.interval_in_ms do
+        {[message], put_in(state, [:pa_messages_last_sent, message.id], now)}
       else
-        {pa_id, {active_pa_message, last_sent_time}}
+        {[], state}
       end
-    end
+    end)
   end
 
-  defp send_pa_message(pa_message) do
-    Enum.each(pa_message.sign_ids, fn sign_id ->
-      send(
-        String.to_existing_atom("Signs/#{sign_id}"),
-        {:play_pa_message, pa_message}
-      )
+  defp play_pa_messages(pa_messages) do
+    Enum.flat_map(pa_messages, fn message ->
+      for sign_id <- message.sign_ids,
+          {sign, should_play?} = GenServer.call(:"Signs/#{sign_id}", {:play_pa_message, message}),
+          should_play? do
+        sign
+      end
```

Comment:
Any thoughts on calling out to all the sign processes concurrently using `Task`?

Reply:
I think using `Task` would be overkill here, since this call should complete more or less instantly.